src/algebra/algebra/spectrum.lean

```diff
@@ -0,0 +1,183 @@
+/-
+Copyright (c) 2021 Jireh Loreaux. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jireh Loreaux
+-/
+import algebra.algebra.basic
+import tactic.noncomm_ring
+/-!
+# Spectrum of an element in an algebra
+This file develops the basic theory of the spectrum of an element of an algebra.
+This theory will serve as the foundation for spectral theory in Banach algebras.
+
+## Main definitions
+
+* `resolvent a`: has type `set R` is the resolvent set of an element `a : A` where
+  `A` is an  `R`-algebra.
+* `spectrum a`: has type `set R` is the spectrum of an element `a : A` where
+  `A` is an  `R`-algebra.
+
+## Main statements
+
+* `unit_left_coset_spectrum`: units in the scalar ring commute (multplication) with the spectrum.
+* `left_add_coset_spectrum`: elements of the scalar ring commute (addition) with the spectrum.
+* `unit_mem_spectrum_mul_iff_swap_mul`: the units (of `R`) in `σ (a*b)` coincide
+  with those in `σ (b*a)`.
+
+## Notations
+
+* `σ a` : `spectrum` of `a : A`
+
+## TODO
+
+* Prove the *spectral mapping theorem* for the polynomial functional calculus
+-/
+
+universes u v
+
+variables {R : Type u} {A : Type v}
+variables [comm_ring R] [ring A] [algebra R A]
+
+
+-- definition and basic properties
+
+/-- Given a commutative ring `R` and an `R`-algebra `A`, the *resolvent* of `a : A`
+is the `set R` consisting of those `r : R` for which `r•1 - a` is a unit of the
+algebra `A`.  -/
+definition resolvent (a : A) : set R :=
+{ r : R | is_unit (r • 1 - a) }
+
+
+/-- Given a commutative ring `R` and an `R`-algebra `A`, the *spectrum* of `a : A`
+is the `set R` consisting of those `r : R` for which `r•1 - a` is not a unit of the
+algebra `A`.
+
+The spectrum is simply the complement of the resolvent.  -/
+definition spectrum (a : A) : set R :=
+(resolvent a)ᶜ
+
+local notation `σ` := spectrum
+
+lemma mem_spectrum_iff_not_unit {r : R} {a : A} :
+  r ∈ (σ a : set R) ↔ ¬ is_unit (r • 1 - a) :=
+by { tidy }
+
+lemma not_mem_spectrum_iff_unit {r : R} {a : A} :
+  r ∉ (σ a : set R) ↔ is_unit (r • 1 - a) :=
+by { apply not_iff_not.mp, simp [set.not_not_mem, mem_spectrum_iff_not_unit] }
+
+lemma mem_resolvent_of_left_right_inverse {r : R} {a b c : A}
+  (h₁ : (r • 1 - a) * b = 1) (h₂ : c * (r • 1 - a) = 1) :
+  r ∈ (resolvent a : set R) :=
+units.is_unit ⟨r•1 - a, b, h₁, by rwa ←left_inv_eq_right_inv h₂ h₁⟩
+
+-- products of scalar units and algebra units
+
+/-- Given a commutative ring `R` and an `R`-algebra `A`, and units `r : units R`
+and `a : units A`, then `unit_mul_unit r a` constructs a `units A` with value `r•a`. -/
+definition unit_mul_unit (r : units R) (a : units A) :
+  units A :=
+⟨r•↑a,
+ r⁻¹•↑a⁻¹,
+ by {simp [smul_mul_smul]},
+ by {simp [smul_mul_smul]}⟩
```

Comment:
```suggestion
```
There is already a smul operation defined on units, so this definition is not needed.